crates/trie/sparse/src/state.rs

```diff
@@ -178,31 +177,35 @@ where
         proof: impl IntoIterator<Item = (Nibbles, Bytes)>,
     ) -> SparseStateTrieResult<()> {
         assert!(!self.retain_updates);
+        // TODO this method doesn't seem to be used, can we remove it?
 
         if self.is_account_revealed(account) {
             return Ok(());
         }
 
-        let mut proof = proof.into_iter().peekable();
+        let proof: Vec<_> = proof.into_iter().collect();
 
-        let Some(root_node) = self.validate_root_node(&mut proof)? else { return Ok(()) };
+        let Some(root_node) = Self::validate_root_node(proof.iter())? else { return Ok(()) };
 
         // Reveal root node if it wasn't already.
         let trie = self.state.reveal_root(root_node, TrieMasks::none(), self.retain_updates)?;
 
-        // Reveal the remaining proof nodes.
+        // Collect the remaining proof nodes to be revealed, skipping the root node and other
+        // previously revealed nodes.
+        let mut nodes_to_reveal = Vec::with_capacity(proof.len() - 1);
         for (path, bytes) in proof {
-            if self.revealed_account_paths.contains(&path) {
+            if path.is_empty() || self.revealed_account_paths.contains(&path) {
                 continue
             }
+
             let node = TrieNode::decode(&mut &bytes[..])?;
-            trie.reveal_node(path, node, TrieMasks::none())?;
+            nodes_to_reveal.push(RevealedSparseNode { path, node, masks: TrieMasks::none() });
 
             // Track the revealed path.
             self.revealed_account_paths.insert(path);
         }
 
-        Ok(())
+        Ok(trie.reveal_nodes(nodes_to_reveal)?)
```

Comment:
can we make `reveal_nodes` accept an iterator instead, so that we can filter and decode the nodes from `proof` without collecting it into a vec?

Reply:
I will try it... we will have to iterate over `proof` once to pull out the root node anyway, but that can be done as a one-off operation separate from filtering/mapping

Reply:
yeah, we can do it the same way as it was before using `peekable`, that will also make the diff smaller

Reply:
Hmmm actually, the first thing that both reveal_nodes implementations need to do is sort the proofs, so they'll need to collect into a Vec anyway. Seems more flexible to do it outside the method.

(also fyi the method this comment is on is one of the ones which isn't used, we should focus on `reveal_decoded_account_multiproof` and `reveal_decoded_storage_multiproof` I think)